### PR TITLE
DOC/BLD: set up automated doc builds with travis / doctr

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes https://github.com/pyproj4/pyproj/issues/256

I added it in such a way that the doc build and upload only happens in a single build (using the DOC=true environment variable)

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Note from `doctr configure` to a repo admin:

```
1. Go to https://github.com/pyproj4/pyproj/settings/keys
   and add the following as a new key:

    ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQC7OqGE1S72vUuOpvA/cv2MYImHu/NgrzBDZYzCzoTekb5nRhkPzBjxsg8TAzm6yM1x1EYiHqNDIqfR/jwEpYvvzkcgrt/I5RjcVNQnVktZe1q/1EkQFblSydnMRFCZMeIyL2ds7MeH4ZgGslA4aUnOc2v5VOMDoLsThVDOjqOBvpm1qktxxxizrSNJgg/MCKweQn0+gOr46/xiBs/bZPfI15oM7c7ug8sr8lSEtzJzGBLFQneAOBMPmxST+LNVP4TwRE29Xh7hhKluC9isanZhZchw7e1591XZegva0lMLszXPTgaT8jRkYXe0RZ//i71RwIaUd5PGswxPYZagc9XS4VoYoLriWXOcVDNc5x2BK7eDr6+7qI/Nk6LALAfrPfmZvZxS/LgZr0M5pq6BGc64/WN2oEwrMUwOJvtcdf8gphVqbg+4GIBrKZkDm54ijLdbGtiZpdzWoNwLTSzphGQMTMUzy94ihqQ9g7nCimK0Qwn2jiC9vFiAaQ8DMttajjQBTXWFhntOOTW3TLkAU3xhrUbMmsohSACxf1QfMQc2gZHUB9ZrkuR4xxiOyAxsZ5iXTDx8lBEzfS6j4qkn2hpS3sOxhBtgjPPqo/H3amF7fV4bRmCL/L4dBb5pmQ4FyDI2U/z8OtxTnJP8HSrXhfm6wwQaWjh4ZF1wSmg/PPWA6Q==
   Be sure to allow write access for the key.
```

See https://drdoctr.github.io/recipes.html#setting-up-doctr-for-a-repo-you-don-t-have-admin-access-to

**Comment by jorisvandenbossche** [comment]
@snowman2 we will only be able to *actually* test it once it is merged in master, since doctr only uploads the docs for commits in master, not in PRs

**Comment by jorisvandenbossche** [comment]
(also changed the sphinx theme, as discussed in the issue)

This should be ready now.

File: .travis.yml [hunk 1]
```diff
@@ -80,6 +83,22 @@ install:
...
   - python -c "import pyproj; pyproj.Proj(init='epsg:4269')"
   - py.test --cov-report term-missing --cov=pyproj -v -s
+  # Building and uploading docs with doctr
+  - set -e
+  - |
+    if [ "$DOC" ]; then
+      pip install sphinx sphinx_rtd_theme
+      cd docs
+      make html
+      cd ..
```
Comment (snowman2) [inline .travis.yml]: What are your thoughts on building the docs outside of the `if` and just be part of the tests?

Also, what are your thoughts on deploying docs in the `after_success` section?

**Comment by jorisvandenbossche** [comment]
> What are your thoughts on building the docs outside of the if and just be part of the tests?

I put it deliberately here, to only build the docs in a single travis build (with the idea: it not needed to build it in each one) 
But for that particular build, it is still "part of the tests", meaning you can check the log output if everything is fine / it will fail if sphinx crashes

> Also, what are your thoughts on deploying docs in the after_success section?

I don't mind much. Looking at the docs, there is a warning about it here: https://drdoctr.github.io/index.html?highlight=after_success#edit-your-travis-file (scroll a bit down). Saying that if you put it in `after_success` it will not cause the build to fail.

**Comment by snowman2** [comment]
Good points. Probably better as is then. Thanks for the clarification!

**Comment by snowman2** [comment]
Thoughts on updating the create_docs.sh and use that for building docs?

**Comment by jorisvandenbossche** [comment]
> Thoughts on updating the create_docs.sh and use that for building docs?

Yep, I actually wanted to propose to move the added code in travis.yml to a separate script. But OK to put that in ci/travis/ then instead of top-level ?

**Comment by snowman2** [comment]
> But OK to put that in ci/travis/ then instead of top-level ?

Are you referring to the entire doctr process or only the sphinx part?

**Comment by jorisvandenbossche** [comment]
Ah, I was referring to the entire process. But you might want to have a script that just builds the docs, as it is now?

**Comment by jorisvandenbossche** [comment]
I am happy to do such a change (update the current `create_docs.sh` for just installing deps + building docs, no doctr) and use that in travis.yml. 
But personally not sure how useful it is to have that in the top-level repo (compared to having a script in travis/ci), as I would assume that somebody locally simply does a `cd docs; make html` ?

**Comment by snowman2** [comment]
It is convenient to have. Much simpler to autocomplete and it is nice have a reproducible way to build docs for newcomers to use as well.

**Comment by jorisvandenbossche** [comment]
I would personally say: then it should not install things (one should try to keep running a script and setting up environment separate IMO)

But, if that's actually the goal, it might be worthwile to add an actual Makefile, so you can eg do `make docs` from the top-level directory. And then we can also provide other development workflows, eg things like `make inplace` as a shortcut for building the C extensions inplace.

**Comment by jorisvandenbossche** [comment]
Anyway, this discussion is getting unrelated to the actual PR :) Will use that script for now, we can always update it in a separate issue/PR.

**Comment by snowman2** [comment]
It would be fantastic to have a Makefile - but that is for another issue as you stated.

**Comment by jorisvandenbossche** [comment]
Updated

File: .travis.yml [hunk 2]
```diff
@@ -80,6 +83,19 @@ install:
 script:
   - python -c "import pyproj; pyproj.Proj(init='epsg:4269')"
   - py.test --cov-report term-missing --cov=pyproj -v -s
+  # Building and uploading docs with doctr
+  - set -e
+  - |
+    if [ "$DOC" ]; then
+      create_docs.sh
```
Comment (snowman2) [inline .travis.yml]: probably need `bash create_docs.sh` here.

**Comment by jorisvandenbossche** [comment]
Note that before merging this, it is best to add the key mentioned above in https://github.com/pyproj4/pyproj/pull/278#issuecomment-485906648

**Comment by jorisvandenbossche** [comment]
@jswhit is this approach fine for you?

**Comment by snowman2** [comment]
Alright, testing now :crossed_fingers:

**Comment by snowman2** [comment]
Looks successful to me. Nice work @jorisvandenbossche!
https://pyproj4.github.io/pyproj/dev/index.html

**Comment by jorisvandenbossche** [comment]
> Looks successful to me.

Great!

**Pull request merged**